src/main/java/org/dice_group/grp/grammar/digram/Digram.java

```diff
@@ -82,4 +78,51 @@ public void setNoOfOccurences(long occurences) {
 		this.occurences = occurences;
 	}
 	
+
+
+	@Override
+	public int hashCode() {
+		final int prime = 31;
+		int result = 1;
+		result = prime * result + ((edgeLabel1 == null) ? 0 : edgeLabel1.hashCode());
+		result = prime * result + ((edgeLabel2 == null) ? 0 : edgeLabel2.hashCode());
+		result = prime * result + ((external == null) ? 0 : external.hashCode());
+		return result;
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (this == obj)
+			return true;
+		if (obj == null)
+			return false;
+		if (getClass() != obj.getClass())
+			return false;
+		Digram other = (Digram) obj;
+		if (edgeLabel1 == null) {
+			if (other.edgeLabel1 != null)
+				return false;
+		} else if (!edgeLabel1.equals(other.edgeLabel1))
+			return false;
+		if (edgeLabel2 == null) {
+			if (other.edgeLabel2 != null)
+				return false;
+		} else if (!edgeLabel2.equals(other.edgeLabel2))
+			return false;
+		if (external == null) {
```

Comment:
externals are not allowed to be empty/null according to the paper. So this case can be removed

Reply:
Well, following the idea of defensive programming, you may simply want to throw an exception with an error message indicating that this case is not allowed :wink:

Reply:
true.

Reply:
my bad, I used the auto-generated equals by the IDE. But I changed it to incorporate your reviews :)